core/src/test/java/io/grpc/internal/AbstractClientStreamTest.java

```diff
@@ -0,0 +1,254 @@
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    stream.inboundPhase(Phase.HEADERS);
+
+    stream.inboundDataReceived(ReadableBuffers.empty());
+
+    verify(mockListener).closed(statusCaptor.capture(), isA(Metadata.class));
+    assertEquals(Code.INTERNAL, statusCaptor.getValue().getCode());
+  }
+
+  @Test
+  public void inboundHeadersReceived_notifiesListener() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+
+    stream.inboundHeadersReceived(headers);
+    verify(mockListener).headersRead(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_failsOnPhaseStatus() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    stream.inboundPhase(Phase.STATUS);
+
+    thrown.expect(IllegalStateException.class);
+
+    stream.inboundHeadersReceived(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_succeedsOnPhaseMessage() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    stream.inboundPhase(Phase.MESSAGE);
+
+    stream.inboundHeadersReceived(headers);
+
+    verify(mockListener).headersRead(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_acceptsGzipEncoding() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    headers.put(ChannelImpl.MESSAGE_ENCODING_KEY, new MessageEncoding.Gzip().getMessageEncoding());
+
+    stream.inboundHeadersReceived(headers);
+    verify(mockListener).headersRead(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_acceptsIdentityEncoding() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    headers.put(ChannelImpl.MESSAGE_ENCODING_KEY, MessageEncoding.NONE.getMessageEncoding());
+
+    stream.inboundHeadersReceived(headers);
+    verify(mockListener).headersRead(headers);
+  }
+
+  @Test
+  public void inboundHeadersReceived_notifiesListenerOnBadEncoding() {
+    AbstractClientStream<Integer> stream =
+        new BaseAbstractClientStream<Integer>(allocator, mockListener);
+    Headers headers = new Metadata.Headers();
+    headers.put(ChannelImpl.MESSAGE_ENCODING_KEY, "bad");
+
+    stream.inboundHeadersReceived(headers);
+
+    verify(mockListener).closed(statusCaptor.capture(), isA(Metadata.class));
+    assertEquals(Code.INVALID_ARGUMENT, statusCaptor.getValue().getCode());
+  }
+
+  /**
+   * No-op base class for testing.
+   */
+  private static class BaseAbstractClientStream<T> extends AbstractClientStream<T> {
```

Comment:
I was using mocks for when I needed to verify calls, and tried to use this when custom behavior was desired.  Ideally this could go into a "testing" directory and could be used by other tests.